validator.go

```diff
@@ -414,7 +418,10 @@ func validateNonEmpty(v interface{}, _ string) error {
 	}
 
 	val := reflect.ValueOf(v)
-	if val.Kind() == reflect.Array || val.Kind() == reflect.Slice {
+	if val.Kind() == reflect.Array || val.Kind() == reflect.Slice || val.Kind() == reflect.Map {
+		if allowNil && val.IsNil() {
+			return nil
+		}
```

Comment:
we might want to return `ErrMissing` instead of `ErrEmpty` if we have a `nil` value.

Reply:
I think it would be weird to return an error that doesn't match the validator type. `nonzero` case is always a `ErrEmpty` everywhere else and required is always `ErrRequired`.

Reply:
I see `validateRequired` will replace the error value to always return `ErrRequired`.

Still the error string is also for the user. Maybe `validateRequired` should still return `ErrEmpty` if the setting is in the config, but does not contain any entry.

Reply:
I recently change validateRequired to do that, a few PR's ago. I assumed that was the behaviour that was expected.

It was surprising to me that the code path returned `ErrEmpty` when the `required` tag was used. But if its more about what is show to the user and less about the developer handling those errors, then I should switch it back.

Reply:
I see. The error type and message should reflect the actual problem that has been found, not the validator that failed. The `required` validator can fail for 2 reasons: a) field was not configured at all b) field was configured, but object or array was empty itself. Failure (a) warrants ErrRequired, while failure (b) should trigger an ErrEmpty error. That is the original `ErrEmpty` for the `required` check was on purpose, not by accident.

ErrRequired is somewhat similar to ErrMissing. The error message says `"missing required field"`. If the array is in the config file but empty the error message should say `"empty field"`. We might even opt to introduce more specific errors like ErrEmptyArray, or ErrEmptyMap, or ErrValueIsZero. WDYT?

Reply:
Yeah I think it would be better to have more detailed errors messages like the `ErrEmptyArray` and the `ErrEmptyMap`. I will work on returning those errors in the correct locations.